### PR TITLE
Tag internal PHP API surface and create index of what is publicly exposed

### LINKED ISSUE #5119: Add @internal phpdoc tags to all classes, functions, and hooks not intended for developers to use
## Feature description

We intend to create auto-generated reference documentation from the codebase from parsing the phpdoc blocks. However, there are many functions, classes, methods, filters, actions, etc. which we do not want developers to use. Either the code in question does not need to be directly invoked by developers, or we may intend to deprecate/replace code as we go through the continued modularization effort. Marking code as `@internal` is also important for the 2.0 version bump where we indicate what we want to commit to supporting.

The classes for sanitizers and embed handlers should be marked as `@internal` except for the base classes which we intend developers to extend.

Certain `@internal` code should also be marked as `@deprecated` eventually.

Adding `@internal` to a class should be sufficient to mark all methods inside also as being `@internal`. The exception to this is filters/actions which may be invoked inside of the class methods. Those should by default not inherit the `@internal` tag and should be explicitly marked as themselves being `@internal` on their own.

See https://github.com/WordPress/phpdoc-parser

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

* 

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

Fixes #5119

This PR adds the necessary `@internal` tags so that phpdoc output can successfully exclude the API surface we do not intend for developers to touch.

When a class has a `@internal` (or `@deprecated`) tag, then it is omitted from the index. If there is a method that does an action or applies a filter, those hooks will be included in the index unless they too have `@internal`. Hooks that have PhpDoc like “This filter is documented in…” are omitted. A couple filters that lacked phpdoc were supplied with it. The `@since` tag was also added in various places.

Functions are included in the index unless they have the `@internal` tag.

Two functions which lack prefixes have been renamed, with the old versions being soft-deprecated (having `@deprecated` tags but no `_deprecated_function()` calls):

* `is_amp_endpoint()` => `amp_is_request()`
* `post_supports_amp()` => `amp_is_post_supported()`

This PR also deprecates the `amp_frontend_show_canonical` filter and the old v0.3 legacy AMP post templates (loaded via the `amp_backcompat_use_v03_templates()` function).

The following table is derived from the following script:

```bash
wp eval-file bin/create-phpdoc-index.php
```

My hope is that the logic in the script can be easily ported for the actual generation of the HTML phpdoc.

Type | Name
-----|-----
action | `amp_customizer_enqueue_preview_scripts`
action | `amp_customizer_enqueue_scripts`
action | `amp_customizer_init`
action | `amp_customizer_register_settings`
action | `amp_customizer_register_ui`
action | `amp_extract_image_dimensions_batch_callbacks_registered`
action | `amp_init`
action | `amp_plugin_update`
action | `amp_post_template_body_open`
action | `amp_post_template_css`
action | `amp_post_template_footer`
action | `amp_post_template_head`
action | `amp_post_template_include_{$template_type}`
action | `amp_print_analytics`
action | `pre_amp_render_post`
class | `AMP_Base_Embed_Handler`
class | `AMP_Base_Sanitizer`
filter | `amp_analytics_entries`
filter | `amp_comment_posted_message`
filter | `amp_content_embed_handlers`
filter | `amp_content_max_width`
filter | `amp_content_sanitizers`
filter | `amp_css_transient_monitoring_sampling_range`
filter | `amp_css_transient_monitoring_threshold`
filter | `amp_customizer_get_settings`
filter | `amp_customizer_is_enabled`
filter | `amp_customizer_post_type`
filter | `amp_dev_mode_element_xpaths`
filter | `amp_dev_mode_enabled`
filter | `amp_enable_optimizer`
filter | `amp_enable_ssr`
filter | `amp_extract_image_dimensions_batch`
filter | `amp_extract_image_dimensions_get_user_agent`
filter | `amp_get_permalink`
filter | `amp_is_enabled`
filter | `amp_mobile_client_side_redirection`
filter | `amp_mobile_user_agents`
filter | `amp_mobile_version_switcher_link_text`
filter | `amp_mobile_version_switcher_styles_used`
filter | `amp_normalized_dimension_extractor_image_url`
filter | `amp_optimizer_config`
filter | `amp_options_menu_is_enabled`
filter | `amp_parsed_css_transient_caching_allowed`
filter | `amp_post_article_footer_meta`
filter | `amp_post_article_header_meta`
filter | `amp_post_status_default_enabled`
filter | `amp_post_template_analytics`
filter | `amp_post_template_customizer_settings`
filter | `amp_post_template_data`
filter | `amp_post_template_dir`
filter | `amp_post_template_file`
filter | `amp_post_template_metadata`
filter | `amp_pre_get_permalink`
filter | `amp_pre_is_mobile`
filter | `amp_query_var`
filter | `amp_reader_themes`
filter | `amp_schemaorg_metadata`
filter | `amp_site_icon_url`
filter | `amp_skip_post`
filter | `amp_supportable_post_types`
filter | `amp_supportable_templates`
filter | `amp_to_amp_excluded_urls`
filter | `amp_to_amp_linking_element_excluded`
filter | `amp_to_amp_linking_enabled`
filter | `amp_validation_error_default_sanitized`
filter | `amp_validation_error_sanitized`
filter | `amp_validation_error_source_file_editor_url_template`
filter | `amp_validation_error_source_file_path`
filter | `amp_validation_error`
function | `amp_add_amphtml_link()`
function | `amp_generate_script_hash()`
function | `amp_get_permalink()`
function | `amp_get_slug()`
function | `amp_is_available()`
function | `amp_is_canonical()`
function | `amp_is_dev_mode()`
function | `amp_is_legacy()`
function | `amp_is_post_supported()`
function | `amp_is_request()`
function | `amp_remove_endpoint()`

## Checklist

- [x] My pull request is addressing an open issue (please create one otherwise).
- [ ] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [ ] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by westonruter** [comment]
Let's add `amp_` prefixes with aliases for `is_amp_endpoint()` and `post_supports_amp()`.

**Comment by westonruter** [comment]
All of the non-deprecated API surface now has `amp_` prefixing.

File: bin/create-phpdoc-index.php [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+	}
+
+	return false;
+}
+
+/**
+ * Process function.
+ *
+ * @param array $parsed_function Parsed function.
+ * @param null  $class           Class name the function belongs to.
```
Comment (pierlon) [inline bin/create-phpdoc-index.php]:
```suggestion
 * @param array         $parsed_function Parsed function.
 * @param string|null  $class            Class name the function belongs to. Defaults to null.
```

**Review by pierlon — CHANGES_REQUESTED** [review]

File: bin/create-phpdoc-index.php (hunk 1 above)
Comment (westonruter) [reply bin/create-phpdoc-index.php]: Aside: This entire file will probably be deleted with the changes that @schlessera is introducing.

**Comment by pierlon** [comment]
@googlebot I consent.

**Comment by pierlon** [comment]
Updated some references of `amp_is_endpoint` and `post_supports_amp` in https://github.com/ampproject/amp-wp/pull/5218/commits/dbe56b44a3cc054410d127aeff2050263fbcb6a1 and https://github.com/ampproject/amp-wp/pull/5218/commits/b1e6575193680130b761af65f1f6d8d177fd8f54.

**Review by pierlon — APPROVED** [review]

File: includes/amp-helper-functions.php [hunk 2]
```diff
@@ -843,9 +860,29 @@ function is_amp_endpoint() {
...
+ *
+ * This function cannot be called before the parse_query action because it needs to be able
+ * to determine the queried object is able to be served as AMP. If 'amp' theme support is not
+ * present, this function returns true just if the query var is present. If theme support is
+ * present, then it returns true in transitional mode if an AMP template is available and the query
+ * var is present, or else in standard mode if just the template is available.
+ *
+ * @since 0.1
+ * @since 2.0 Renamed to AMP-prefixed version, amp_is_endpoint().
+ * @deprecated Use amp_is_endpoint() instead.
```
Comment (schlessera) [inline includes/amp-helper-functions.php:897]: `Use amp_is_endpoint() instead.` => `Use amp_is_request() instead.`

Comment (schlessera) [reply includes/amp-helper-functions.php:897]: I'll send in a quick PR for this.

Comment (schlessera) [reply includes/amp-helper-functions.php:897]: See https://github.com/ampproject/amp-wp/pull/5231

**Pull request merged**